resource/graph_node_test.go

```diff
@@ -147,26 +153,55 @@ func lifecycleTest(t *testing.T, node *resource.GraphNode, initialDeps []string)
 	test.That(t, err, test.ShouldBeNil)
 	test.That(t, res, test.ShouldEqual, ourRes2)
 	test.That(t, node.NeedsReconfigure(), test.ShouldBeTrue)
+	test.That(t, node.ShouldTryReconfigure(), test.ShouldBeTrue)
 	test.That(t, node.Config(), test.ShouldResemble, resource.Config{Attributes: utils.AttributeMap{"1": 2}})
 	test.That(t, node.UnresolvedDependencies(), test.ShouldResemble, []string{"3", "4", "5"})
 
-	// but an error happened
+	// an error happens 5 (maxReconfigAttempts) times
+	for i := 0; i < 5; i++ {
+		node.IncrementTimesReconfigured()
+		node.SetLastError(ourErr)
+		_, err = node.Resource()
+		test.That(t, err, test.ShouldNotBeNil)
+		test.That(t, err.Error(), test.ShouldContainSubstring, ourErr.Error())
+		res, err = node.UnsafeResource()
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, res, test.ShouldEqual, ourRes2)
+		test.That(t, node.IsUninitialized(), test.ShouldBeFalse)
+	}
+	test.That(t, node.ShouldTryReconfigure(), test.ShouldBeFalse)
+
+	// retry with new config
+	ourConf = resource.Config{Attributes: utils.AttributeMap{"1": 2}}
+	node.SetNewConfig(ourConf, []string{"6", "7", "8"})
```

Comment:
This should simulate the user changing the config and resetting the # of attempts. I think the method that invokes `SetNewConfig` is called when the resource is marked for update in `updateResources` when the diff between the old config and new config is evaluated.

Reply:
>  I think the method that invokes SetNewConfig to kick off the reset is called when the resource is marked for update in updateResources (when the diff between the old config and new config is evaluated).

Correct! I think this test is looking good; within `graph_node_test`, the concept of "reconfiguring" a resource could be represented by `SetNewConfig`, `SetNeedsUpdate` and `SwapResource`. If you could add test cases for `SetNeedsUpdate` and `SwapResource` resetting the reconfigure attempt count, I think we'd be covered.